### PR TITLE
feat(multi-tenancy): add support for multi-tenancy in bulk loader

### PR DESCRIPTION
**Pull request by NamanJain8**
This PR adds bulk loader support in multi-tenancy.

Adds a flag --force-namespace in bulk loader which makes it load the data into a specified namespace. If this flag is not set, then it preserves the namespace from the schema file.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7399)

### CODE REVIEW

File: chunker/rdf_parser.go
Comment (NamanJain8) [inline chunker/rdf_parser.go:236]: Do we need a separate field in NQuad just for the namespace? We can use the label field?
The label is `string` and Namespace is `uint64`.

Comment (NamanJain8) [reply chunker/rdf_parser.go:236]: In the current implementation, the user cannot pass non-namespace strings in the label.
We can allow it in the following way:
The user tells that his RDFs contain a label and also tell us to load the data into a specific namespace. Then while parsing the RDFs we will not treat the label as namespace.

**Review by manishrjain — APPROVED** [review]
:lgtm: Whole bunch of TODOs, so let's resolve them before merging. Leave it to @jarifibrahim for final approval.

> Reviewed 16 of 17 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-:-MSmsmfk2f7UsRJsApZR:bofw9ck)* status: all files reviewed, 7 unresolved discussions (waiting on @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[chunker/json_parser.go, line 464 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmpfz65jP20WWi03eJ:-MSmpfz65jP20WWi03eK:b4t63tj) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/chunker/json_parser.go#L464)):*
> ```Go
> 		// this int64 case is needed for FastParseJSON, which doesn't use json.Number
> 		case int64:
> 			namespace = uint64(ns)
> ```

Don't make it a number in JSON. Use hex strings. Add a string part. Do strip spaces and use strconv.ParseUint.
___
*[chunker/rdf_parser.go, line 231 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmqESE0YCgCxmRSQa5:-MSmqESE0YCgCxmRSQa6:b-lrjy4b) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/chunker/rdf_parser.go#L231)):*
> ```Go
> 	// TODO(Naman): Ensure the label contains valid namespace.
> 	// Append the namespace to the predicate before returning NQuad.
> 	if rnq.Label != "" {
> ```

why not rename label to namespace?
___
*[chunker/rdf_parser.go, line 232 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmq9Az4nqgHzg_Q03y:-MSmq9Az4nqgHzg_Q03z:bnjg7gp) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/chunker/rdf_parser.go#L232)):*
> ```Go
> 	// Append the namespace to the predicate before returning NQuad.
> 	if rnq.Label != "" {
> 		ns, err := strconv.ParseUint(rnq.Label, 0, 64)
> ```

do strip spaces.
___
*[dgraph/cmd/bulk/merge_shards.go, line 63 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmrt3o1Zx4ZEnsaD42:-MSmrt3pCozP1JnSEqLI:bcdui1p) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/dgraph/cmd/bulk/merge_shards.go#L63)):*
> ```Go
> 	// Put the first map shard in the first reduce shard since it contains all the reserved
> 	// predicates.
> 	// TODO(Naman): Why do we handle first shard differently?
> ```

This is because we want all reserved stuff to be in group 1. You can add this as a comment.
___
*[dgraph/cmd/bulk/schema.go, line 92 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmsIaq2ki8XQ4TwbB0:-MSmsIaq2ki8XQ4TwbB1:bmnc6n8) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/dgraph/cmd/bulk/schema.go#L92)):*
> ```Go
> // checkAndSetInitialSchema initializes the schema for namespace if it does not already exist.
> func (s *schemaStore) checkAndSetInitialSchema(namespace uint64) {
> 	if _, ok := s.namespaces.Load(namespace); !ok {
> ```

if ok; return

lock

if ok return

do stuff.
___
*[dgraph/cmd/debug/run.go, line 574 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmsPII79qqElTm84F9:-MSmsPII79qqElTm84FA:b-ye83c3) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/dgraph/cmd/debug/run.go#L574)):*
> ```Go
> 		}
> 		ns, attr := x.ParseNamespaceAttr(pk.Attr)
> 		x.Check2(buf.WriteString(fmt.Sprintf(" ns: 0x%x ", ns)))
> ```

%#x would give you the 0x.

**Review by jarifibrahim — APPROVED** [review]

**Review by jarifibrahim — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-:-MT-ZQiG380ZXOuNiLpw:bnfp4nl)* status: 7 of 15 files reviewed, 11 unresolved discussions (waiting on @manishrjain, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[go.mod, line 5 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MT-SsYA1TbKiwubRRHq:-MT-SsYA1TbKiwubRRHr:b-cmooqt) ([raw file](https://github.com/dgraph-io/dgraph/blob/90b9768c22f523cf2755cae3dbb4bf8ebc0901a1/go.mod#L5)):*
> ```XML
> go 1.12
> 
> replace github.com/dgraph-io/dgo/v200 => /home/algod/go/src/github.com/dgraph-io/dgo
> ```

Remove this, please.
___
*[dgraph/cmd/bulk/loader.go, line 173 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MT-U_u61_mSZr8rDMhP:-MT-U_u61_mSZr8rDMhQ:b-b411ym) ([raw file](https://github.com/dgraph-io/dgraph/blob/90b9768c22f523cf2755cae3dbb4bf8ebc0901a1/dgraph/cmd/bulk/loader.go#L173)):*
> ```Go
> func (ld *loader) leaseNamespaces() {
> 	var maxNs uint64
> 	ld.namespaces.Range(func(key, value interface{}) bool {
> ```

Add a comment.
___
*[dgraph/cmd/bulk/loader.go, line 290 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MT-Y09d3fkgU9FB4b6F:-MT-Y09e0HADA7pce58h:b-oivtsl) ([raw file](https://github.com/dgraph-io/dgraph/blob/90b9768c22f523cf2755cae3dbb4bf8ebc0901a1/dgraph/cmd/bulk/loader.go#L290)):*
> ```Go
> 
> 	// Send the graphql triples
> 	// TODO(Naman): Handle this.
> ```

Put more details here.
___
*[worker/export_test.go, line 76 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MT-ZFQnCS8zDI0ZiTHB:-MT-ZFQnCS8zDI0ZiTHC:b-xp3rxi) ([raw file](https://github.com/dgraph-io/dgraph/blob/597891dfde1afcc9f37c1ca1e39ec4b0ef7921c4/worker/export_test.go#L76)):*
> ```Go
> func populateGraphExport(t *testing.T) {
> 	rdfEdges := []string{
> 		`<1> <friend> <5> <author0> .`,
> ```

Add namespace over here and then validate it in a test.

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7399)* status: 3 of 26 files reviewed, 11 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[go.mod, line 5 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MT-SsYA1TbKiwubRRHq:-MT0bvOIARybTaXqdQ5k:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/90b9768c22f523cf2755cae3dbb4bf8ebc0901a1/go.mod#L5)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Remove this, please.
</blockquote></details>

Done.
___
*[chunker/json_parser.go, line 464 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmpfz65jP20WWi03eJ:-MT0c1fjCMkHOOfE9xOv:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/chunker/json_parser.go#L464)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't make it a number in JSON. Use hex strings. Add a string part. Do strip spaces and use strconv.ParseUint.
</blockquote></details>

Done.
___
*[chunker/rdf_parser.go, line 231 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmqESE0YCgCxmRSQa5:-MT0c6-QAjCV1PydKDKY:bb1wzn6) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/chunker/rdf_parser.go#L231)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

why not rename label to namespace?
</blockquote></details>

We have added a new field `Namespace` of type `uint64` in the Nquad. If we had used `string` type, we will have to do `strconv.ParseUint(ns...)` at a bunch of places where we need namespace for the same Nquad.
Moreover, the `Label` is not used by dgraph and just adds up an extra field in `pb.Posting`.  
___
*[chunker/rdf_parser.go, line 232 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmq9Az4nqgHzg_Q03y:-MT0c5-V6gi4SdHy2OO5:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/chunker/rdf_parser.go#L232)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

do strip spaces.
</blockquote></details>

Done.
___
*[dgraph/cmd/bulk/loader.go, line 173 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MT-U_u61_mSZr8rDMhP:-MT0cvrdCE0_b5_ofII8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/90b9768c22f523cf2755cae3dbb4bf8ebc0901a1/dgraph/cmd/bulk/loader.go#L173)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Add a comment.
</blockquote></details>

Done.
___
*[dgraph/cmd/bulk/loader.go, line 290 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MT-Y09d3fkgU9FB4b6F:-MT0cwor7kZkGkzDKZyj:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/90b9768c22f523cf2755cae3dbb4bf8ebc0901a1/dgraph/cmd/bulk/loader.go#L290)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Put more details here.
</blockquote></details>

Done.
___
*[dgraph/cmd/bulk/merge_shards.go, line 63 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmrt3o1Zx4ZEnsaD42:-MT0cxim4xJxpyM2I1Yx:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/dgraph/cmd/bulk/merge_shards.go#L63)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This is because we want all reserved stuff to be in group 1. You can add this as a comment.
</blockquote></details>

Done.
___
*[dgraph/cmd/bulk/schema.go, line 92 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmsIaq2ki8XQ4TwbB0:-MT0cyN618pZO7J_mfT1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/dgraph/cmd/bulk/schema.go#L92)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

if ok; return

lock

if ok return

do stuff.
</blockquote></details>

Done.
___
*[dgraph/cmd/debug/run.go, line 574 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7399#-MSmsPII79qqElTm84F9:-MT0czTUCf2xTIGbv0Pn:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/59e71787137634d316844b036f5ddbcada3721d8/dgraph/cmd/debug/run.go#L574)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

%#x would give you the 0x.
</blockquote></details>

Done.

**Pull request merged**
